### ISSUE TITLE
Base image of Azul image is not pinned to digest

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#6446]
The image reference is currently using the image name and the tag. Tags are mutable, so the base image could change.

This issue caused a failing build https://ucsc-gi.slack.com/archives/C705Y6G9Z/p1721860880150469.

The runner image Dockerfile has the same issue.

**Comment by hannes-ucsc** [comment]
```diff
Index: src/azul/docker.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/docker.py b/src/azul/docker.py
--- a/src/azul/docker.py	(revision c6dbdb4b7d06ed7f6bbc38afc6b5c7f6f95cc60f)
+++ b/src/azul/docker.py	(date 1721875654896)
@@ -48,6 +48,7 @@
     config,
     require,
 )
+from azul.deployment import aws
 from azul.types import (
     JSONs,
 )
@@ -458,6 +459,11 @@
             username, password = self._docker_io_auth
             dxf.authenticate(username=username,
                              password=password,
+                             response=response)
+        elif host.endswith(f'ecr.{config.region}.amazonaws.com'):
+            username, password = self._ecr_auth
+            dxf.authenticate(username='AWS',
+                             password='eyJwYXlsb2FkIjoiakJKcnQyS2p6ZlEveDQ2R0NPZHdHS3EzUitTNjBQT3ArZkEwL2lMNHZ3QWJxbnZaSXRZNHRWRm5qYk1CSnNuenpDMVV0ak5KSC9KNkUvdGxoazgwSTJjVkdLdEZubTBFc1ZuV2w0aHl1UCtUV1k4UTB3LzNDVFpIV2wySFpqLzJqdEV3Z1hnZUpFMFJPVE1jaFQ0bUo3RklVTGdxRHFCWUhQempoUEN2ckdDLzVYeEJHVEMxNEVXNGRMNnRhV0szS0xmTVhua2ZlZ0dYcktwZitSejFkQURSaG44bnZBV0F3TVJCNDRvcnpmdVhzd21ZZVRYWTNGclJxWktoQWZNdDZpY3AxWk0vbHZNL0ZUL2l4N3BKaXBRSXM1ZjRmbzUvbDdKd1RvbHRCN2VaZ25wd1FVcVJxZHFUMUZVZlF1K21sMnoyM2ttdW92Z2RMYXN2aXNHbis0RWExSnh1TWttQjl5SWFLOC9xSm5YdjZ2Slh1ZTIrVzVVY01Tc2hzL2FXTGttZGZ4ZFRwK1ZGeERidnZpc1NCcVdLNGdnVlFjREptVFZJU2pkNjVLTFU5WVBWdU5OVUxrSytLTldNNGF2Zk4wMnZxcm85YzJkOTZxc2xFYTUxaDNnb292eUdUYmozQ1dhQmFZYSszVGlRc2VnVDFYQUxxb2sxbkxLc1d3MHh5UVVJWkRQQnhYWkhpNDkybXVWRG04cHNHN3oyM3RIcXlTMjBKWjNCaTJxQ0s2bXI3OXhNeEgreVM0MFZPOXZ3ck93VElNQS9YSTU1ejk1ci9RSks4QUJsMkwvcGhGbEpPWnZNTXRYbUVhS3prYlFWMytpNTFWV0h2NkRndTkxNjd2Si95dWorZFRJT0Y1dFIzK2lSMURtdXVjNXpOYnFsaUF1QjhvdDduaHh4K2x0Sy9VeU5NdDB3Q3V0eW9pLzhIOCsxc1lEbk5QMThMQ2c1RDlZYmgxMk1tWUlLVWF2YTVmbVVNZ1Q2MENreXc3QWR6UDJGVm5yTWJYQzkvd053VWZKWTVvTEpVVWYvalJ0eWdCZlc0MW9iYWFHMzNTUkRoVkNHazN5cUk2S1BQTlZSQXgrT2lVVlVrOUFLZ0ZSclNkTlphRVl2ejliOERIMytjMGx6anlYR2UyOG9NbERIUmJzVWlBVENjd0syT1hUa2pmQlF1Z3pIRFFDckgvcndqSWEwdlVIemZFbDZJczFRRzhrQzVtV1VBOFRaOGRmVStQSXA4WUlRbGJVYktkdFpvNklqMlV0bnZxb0doekpDWGcwNkdRbkE4b0k0TUlmRngxdE9DSmhBZFlMQ2dGMm1FclNpZm5uU2hXV1RKTXJTaDFtZHdwSEFqZ2dXaUphd2NQUkJsNlhRbERSaHNYQnRZNVhraVpWSDBYZWdTaVY2SFpMOTRQaWJDUVF3bjVUbmVXYnJvZ21lRm1lbUZiZWxIK3QyUHZ5ZzFxTTNmQUJ2Rnl0ZmFaZUp0UTdpUm4xcm5GRFh3cEdpVEtUOFBVWUxpVWVzNWxkejNKOXdiSU50VVE0OTJIbGQ1RytvM3pDUkwxazRQTDJWUUZLcU1BL2gwdStJVThLeEhxOFU0Tjk2N0cyQUZGWkx3Ry9GbmNBb0ZIdUxHN0xEK2FpMC8yRkF4VmFUc25lWXMvYUhQTklrTjZEZmVnbnY4cE0zVHh6MXlnPT0iLCJkYXRha2V5IjoiQVFFQkFIaHdtMFlhSVNKZVJ0Sm01bjFHNnVxZWVrWHVvWFhQZTVVRmNlOVJxOC8xNHdBQUFINHdmQVlKS29aSWh2Y05BUWNHb0c4d2JRSUJBREJvQmdrcWhraUc5dzBCQndFd0hnWUpZSVpJQVdVREJBRXVNQkVFRFBCYlNuaG12cE05Uk1Ja1ZBSUJFSUE3YnArcFlKN2xKV3U3N21PaDRUZDVzRkhNTTdJWDY4RVByYWtDRTRhcmIrSXQ0MDhhZ1d2RVhJVzRWcG52am5zdjhoMnIrVW0yRkRjbzRMQT0iLCJ2ZXJzaW9uIjoiMiIsInR5cGUiOiJEQVRBX0tFWSIsImV4cGlyYXRpb24iOjE3MjE5MTY1NzN9',
                              response=response)
         else:
             raise NotImplementedError(host)
@@ -476,6 +482,11 @@
         output = json.loads(output)
         return output['Username'], output['Secret']
 
+    @cached_property
+    def _ecr_auth(self) -> tuple[str, str]:
+        response = aws.ecr.get_authorization_token()
+        return 'AWS', one(response['authorizationData'])['authorizationToken']
+
 
 def pull_docker_image(ref: ImageRef) -> Image:
     return _push_or_pull(ref, 'pull')
```

**Comment by hannes-ucsc** [comment]
With this patch

```diff
Index: src/azul/docker.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/docker.py b/src/azul/docker.py
--- a/src/azul/docker.py    (revision c6dbdb4b7d06ed7f6bbc38afc6b5c7f6f95cc60f)
+++ b/src/azul/docker.py    (date 1721887215820)
@@ -2,6 +2,7 @@
     ABCMeta,
     abstractmethod,
 )
+import base64
 from base64 import (
     urlsafe_b64encode,
 )
@@ -48,6 +49,9 @@
     config,
     require,
 )
+from azul.deployment import (
+    aws,
+)
 from azul.types import (
     JSONs,
 )
@@ -456,6 +460,11 @@
                              response=response)
         elif host.endswith('.docker.io') or host == 'docker.io':
             username, password = self._docker_io_auth
+            dxf.authenticate(username=username,
+                             password=password,
+                             response=response)
+        elif host.endswith(f'ecr.{config.region}.amazonaws.com'):
+            username, password = self._ecr_auth
             dxf.authenticate(username=username,
                              password=password,
                              response=response)
@@ -476,6 +485,14 @@
         output = json.loads(output)
         return output['Username'], output['Secret']
 
+    @cached_property
+    def _ecr_auth(self) -> tuple[str, str]:
+        response = aws.ecr.get_authorization_token()
+        token = one(response['authorizationData'])['authorizationToken']
+        token = base64.b64decode(token).decode('ascii')
+        username, _, password = token.partition(':')
+        return username, password
+
 
 def pull_docker_image(ref: ImageRef) -> Image:
     return _push_or_pull(ref, 'pull')
```

we can use the Docker registry API endpoint of ECR to determine the repository digest and manifest list, but, more importantly, also the image ID of each image in the list:

```python
>>> from azul.docker import Repository
>>> from azul import config
>>> r = Repository(host=config.docker_registry[:-1], name='docker.io/library/python')
>>> digest = r.get_tag('3.11.9-bullseye')
>>> digest
'sha256:0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9'
>>> r.get_manifest(digest)
```

This produces `{'digest': 'sha256:0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9', 'manifests': {'linux/arm64': {'digest': 'sha256:da5621baeccf23575f1dc7b71af2521c328d0826ec7e7e9726208e695ac53e67', 'id': 'sha256:addfe610521b3a1560e1303dcf377c3f2509cdbab82e5694845c502c202ccc14', 'platform': 'linux/arm64'}, 'linux/amd64': {'digest': 'sha256:df5d09c4fcabb63cfa30ac8458268039ba2136563bd59b7b429da1f1b041ac97', 'id': 'sha256:b17d2a63762774a0d91afb5aa4a5621804c09d1bb4343dfc15e607161e03d28a', 'platform': 'linux/amd64'}}}`

Unlike the ECR API DescribeImage response, this contains the image ID for each platform-specific image. That ID is universal and can be matched against `image_manifests.json`. After ensuring that ECR has the expected image, we could then refer to the Dockerfile base image using the digest:

```diff
Index: Makefile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Makefile b/Makefile
--- a/Makefile	(revision c6dbdb4b7d06ed7f6bbc38afc6b5c7f6f95cc60f)
+++ b/Makefile	(date 1721872044022)
@@ -38,7 +38,7 @@
 docker$1: check_docker
 	docker build \
 	       --build-arg azul_docker_registry=$$(azul_docker_registry) \
-	       --build-arg azul_python_image=$$(azul_python_image) \
+	       --build-arg azul_python_image=docker.io/library/python@sha256:0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9 \
 	       --build-arg azul_docker_version=$$(azul_docker_version) \
 	       --build-arg azul_terraform_version=$$(azul_terraform_version) \
 	       --build-arg PIP_DISABLE_PIP_VERSION_CHECK=$$(PIP_DISABLE_PIP_VERSION_CHECK) \
```

Here we hard-coded the digest, but we could easily resolve it dynamically, similar to how we resolve images for pulling or launching. But that requires Python and a fully populated venv. Unfortunately, the `build_image` job on GitLab is running on the bare-bones runner image, which doesn't have Python installed.

We could use the approach for the runner image (since it is built locally, from a fully configured and provisioned working copy), but for the Azul image this seems like a dead end.

**Comment by hannes-ucsc** [comment]
ECR does actually expose the image ID

```
$ aws ecr describe-images --repository-name docker.io/library/python
{
    "imageDetails": [
        {
            "registryId": "122796619775",
            "repositoryName": "docker.io/library/python",
            "imageDigest": "sha256:df5d09c4fcabb63cfa30ac8458268039ba2136563bd59b7b429da1f1b041ac97",
            "imageTags": [
                "3.11.9-bullseye-linux-amd64"
            ],
            "imageSizeInBytes": 351743963,
            "imagePushedAt": 1719987825.0,
            "imageManifestMediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "artifactMediaType": "application/vnd.docker.container.image.v1+json",
            "lastRecordedPullTime": 1721874035.925
        },
        {
            "registryId": "122796619775",
            "repositoryName": "docker.io/library/python",
            "imageDigest": "sha256:0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9",
            "imageTags": [
                "3.11.9-bullseye"
            ],
            "imageSizeInBytes": 351743963,
            "imagePushedAt": 1719987854.0,
            "imageManifestMediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
            "lastRecordedPullTime": 1721916198.45
        },
        {
            "registryId": "122796619775",
            "repositoryName": "docker.io/library/python",
            "imageDigest": "sha256:da5621baeccf23575f1dc7b71af2521c328d0826ec7e7e9726208e695ac53e67",
            "imageTags": [
                "3.11.9-bullseye-linux-arm64"
            ],
            "imageSizeInBytes": 343417502,
            "imagePushedAt": 1719987850.0,
            "imageManifestMediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "artifactMediaType": "application/vnd.docker.container.image.v1+json",
            "lastRecordedPullTime": 1721925998.455
        }
    ]
}
```

For the multi-platform image, we can get the manifest list:

```
$ aws ecr batch-get-image --repository-name docker.io/library/python --image-ids imageDigest=sha256:0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9 | jq -r '.images[0].imageManifest' | jq
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 2007,
      "digest": "sha256:da5621baeccf23575f1dc7b71af2521c328d0826ec7e7e9726208e695ac53e67",
      "platform": {
        "architecture": "arm64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 2007,
      "digest": "sha256:df5d09c4fcabb63cfa30ac8458268039ba2136563bd59b7b429da1f1b041ac97",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    }
  ]
}
```

For the platform-specific images that are part of the list we can get the universal, i.e., portable image ID

```
$ aws ecr batch-get-image --repository-name docker.io/library/python --image-ids imageDigest=sha256:df5d09c4fcabb63cfa30ac8458268039ba2136563bd59b7b429da1f1b041ac97 | jq -r '.images[0].imageManifest|fromjson|.config.digest'
sha256:b17d2a63762774a0d91afb5aa4a5621804c09d1bb4343dfc15e607161e03d28a
```

The ID matches that produced by the approach outlined in the previous comment.

**Comment by hannes-ucsc** [comment]
The ECR image digest is the SHA256 of the image manifest. This is true for both, the image lists (multi-platform images) and the platform-specific images contained therein:

```
$ aws ecr batch-get-image --repository-name docker.io/library/python --image-ids imageDigest=sha256:df5d09c4fcabb63cfa30ac8458268039ba2136563bd59b7b429da1f1b041ac97 | jq -r '.images[0].imageManifest' | ghead -c -1 | sha256sum
df5d09c4fcabb63cfa30ac8458268039ba2136563bd59b7b429da1f1b041ac97  -
(.venv) hannes-m1bp:azul hannes$ aws ecr batch-get-image --repository-name docker.io/library/python --image-ids imageDigest=sha256:0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9 | jq -r '.images[0].imageManifest' | ghead -c -1 | sha256sum
0aaade21a262476ce47e7a80b706bd853754078e3b63b7754ca9f61935db86f9  -
```

The image digests are portable across different ECR registries in different accounts:

<img width="2589" alt="image" src="https://github.com/user-attachments/assets/603db8c4-5b6d-4f43-8ad6-f36728986545">

**Comment by hannes-ucsc** [comment]
Here is a diff between a manifest from docker.io (left) and that from ECR (right):

<img width="1616" alt="image" src="https://github.com/user-attachments/assets/85293ea2-112c-402e-a6b3-c1bade60aa4f">

The left side was obtained with `DXF.get_image_manifest_and_response`, the right side with `aws ecr batch-get-image`.

Aside from the different media types, there are whitespace and property ordering differences. It's a bit problematic that the digest is sensitive to what should be insignificant differences.

Nevertheless, we should be able to predict the ECR manifest by transforming the docker.io manifest to the ECR form and then calculating the SHA256 of the result to produce the ECR digest of the single-platform image. The list manifest can be generated from the individual digests and the result can be hashed to produce the ECR digest of the multi-platform image.

I am uncertain who or what produces these manifests, if it is the `docker` client pushing them to ECR, or if ECR makes them. The risk is that prediction breaks all the time after, say, changes to the `docker` client version, or when ECR updates their algorithm. As I wrote above, the ECR digest are currently consistent across accounts but if ECR staggers the roll-out of a change to the manifest format, we'd be screwed. To evaluate that risk, we should implement the digest prediction soon, track the result in `image_manifests.json` and compare the predicted digests to the actual ones for a while. Unless the prediction is too brittle, we can then simplify the image resolution in `docker.py` (`resolve_docker_image_for_launch` and `…_pull`) and `make environment.boot` could emit the predicted digest as part of `azul_python_image` to form `image@sha256:…` reference rather than a `image:tag` one.

**Comment by hannes-ucsc** [comment]
Unfortunately, because we currently use `docker pull` followed by `docker push` for copying images, which involves decompression and recompression of the layers, some of the layers on ECR end up with a different digest than on docker.io. The ECR digest is much more difficult to predict than I thought, and probably very brittle.

However, I discovered that `skopeo` doesn't involve compression/decompression so we might get away with using that for copying instead. I did a test and the repo digest on ECR was the same as that on docker.io. This is great news because it simplifies image resolution even further. We would still need to predict and track the ECR digests of multi-platform images because they may be a subset of those on docker.io. I think I have a good grasp on what this would entail.

Skopeo can use credentials managed by the `docker` client (or is it the daemon?) as long as [credStore](https://github.com/containers/skopeo/issues/1083) isn't used. It is the default on Docker Desktop so macOS users would have to remove it and accept the fact that their Docker credentials are stored in plaintext on their machines. For ECR these credentials are short-lived tokens. Only the PAT for DockerHub is long-lived.
